datumaro/components/media.py

```diff
@@ -178,3 +179,255 @@ def save(self, path):
+        if not self._includes_frame(idx):
+            raise IndexError(f"Video doesn't contain frame #{idx}.")
+
+        return self._get_iterator()[idx]
+
+    def get_frame_data(self, idx: int) -> VideoFrame:
+        if not self._includes_frame(idx):
+            raise IndexError(f"Video doesn't contain frame #{idx}.")
+
+        return self._get_iterator().get_frame_data(idx)
+
+    def __iter__(self) -> Iterator[VideoFrame]:
+        """
+        Iterates over frames lazily, if possible.
+        """
+
+        if self._frame_count is not None:
+            # Decoding is not necessary to get frame pointers
+            # However, it can be inacurrate
+            end_frame = self._get_end_frame()
+            for index in range(self._start_frame, end_frame, self._step):
+                yield VideoFrame(video=self, index=index)
+        else:
+            # Need to decode to iterate over frames
+            yield from self._get_iterator()
+
+    @property
+    def frame_count(self) -> Optional[int]:
+        """
+        Returns frame count, if video provides such information.
+        Note that not all videos provide length / duration metainfo.
+
+        The count is affected by the frame filtering options of the object.
+        """
+        return self._get_length()
+
+    @property
+    def frame_size(self) -> Tuple[int, int]:
+        if self._frame_size is None:
+            self._frame_size = self._get_frame_size()
+        return self._frame_size
+
+    def _get_frame_size(self) -> Tuple[int, int]:
+        w = self._container.get(cv2.CAP_PROP_FRAME_WIDTH)
+        h = self._container.get(cv2.CAP_PROP_FRAME_HEIGHT)
+
+        if h and w:
+            frame_size = (int(h), int(w))
+        else:
+            image = next(self._get_iterator()).data
+            frame_size = image.shape[0:2]
+
+        return frame_size
+
+    @staticmethod
+    def _get_frame_count(container) -> Optional[int]:
+        # Not all videos provide length / duration metainfo
+        # Note that this information can be invalid or inaccurate
+        # due to variable frame rate
+        # https://stackoverflow.com/a/47796468
+        video_length = container.get(cv2.CAP_PROP_FRAME_COUNT)
+        if video_length:
+            video_length = int(video_length)
+        else:
+            video_length = None
+        return video_length
+
+    def _get_end_frame(self):
+        if self._end_frame is not None and self._frame_count is not None:
+            end_frame = min(self._end_frame, self._frame_count)
+        else:
+            end_frame = self._end_frame or self._frame_count
+
+        return end_frame
+
+    def _get_length(self):
+        if self._length is None:
+            end_frame = self._get_end_frame()
+
+            length = None
+            if end_frame is not None:
+                length = (end_frame - self._start_frame) // self._step
+                assert 0 < length
+
+            self._length = length
+
+        return self._length
```

Comment:
This seems like a good fit for a `functools.cached_property` (or maybe just a regular property - it only requires a few arithmetic operations to compute).

Reply:
The value is already cached. I moved the code to the existing property for this.

Reply:
Okay, fair enough.